### ISSUE TITLE
[REQ] Support for OpenAPI Spec 3.1

### ISSUE THREAD

**Issue by westn** [OpenAPITools/openapi-generator#9083]
### Is your feature request related to a problem? Please describe.

Support for OpenAPI Spec 3.1, released last month: https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0

## Describe the solution you'd like
Add support for OpenAPI Spec 3.1

## Additional context
Couldn't see any other mention of this besides a short mention in #15 
Another note on https://www.openapis.org/blog/2020/06/18/openapi-3-1-0-rc0-its-here which mention that the specification-change is bigger than it sounds and might break some tooling.

**Comment by spacether** [comment]
We use [swagger-parser](https://github.com/OpenAPITools/openapi-generator/blob/master/pom.xml#L1542) to parse openapi specs. [That project must implement v 3.1 spec parsing](https://github.com/swagger-api/swagger-parser/issues/1535) before we can implement it in the generators.

**Comment by darkbasic** [comment]
https://github.com/swagger-api/swagger-parser/commits/oas3.1

**Comment by cgfarmer4** [comment]
Any updates to share on potentially updating the generators to OAS 3.1?

**Comment by gunsluo** [comment]
> Any updates to share on potentially updating the generators to OAS 3.1?

https://github.com/swagger-api/swagger-parser/issues/1535,  there has been no update for half a year.

**Comment by cyangle** [comment]
OAS 3.1 support has been added to swagger-parser in https://github.com/swagger-api/swagger-parser/pull/1730

**Comment by denisgolubovic** [comment]
What release do you aim for this to be implemented in the generator? 6.1.0?

**Comment by kiwisincebirth** [comment]
swagger-parser 2.1 has been released with OAS 3.1 support
https://github.com/swagger-api/swagger-parser/releases/tag/v2.1.0

**Comment by RickKukiela** [comment]
This is a big oof on my part. I just spent a ton of time setting up an API spec for a client and used 3.1.0 and now that its all approved I need to actually program the API server and need to generate the base files from the spec. Now I find out that not only does this package not support 3.1.0 currently but I cannot seem to find any code generator that does :( big yikes! I have to figure out what my options are at this point as my spec currently makes heavy use of the `type` as an `array` not a `string` and I also use `oneOf` with `"null"` or a `$ref` throughout the whole thing which makes it a major problem to downgrade to `3.0.3` at this point.

I see that as of `27 days ago` at the time of me writing this, the 3.1 spec was added to https://github.com/swagger-api/swagger-parser/issues/1535 - Does anyone have any idea when there will be a beta / snapshot release that will support 3.1.0 as if its going to be soon(tm), I might keep just waiting for that as an option before nuking my 3.1.0 setup....

**Comment by slPerryRhodan** [comment]
We are also waiting for this feature. I hope there will come soon a new release for that.

**Comment by Stokestack** [comment]
This is a huge impediment. I too have an API that relies on 3.1 features. It seems to be taking the toolchain a very long time to support it.

**Comment by spacether** [comment]
> This is a big oof on my part. I just spent a ton of time setting up an API spec for a client and used 3.1.0 and now that its all approved I need to actually program the API server and need to generate the base files from the spec. Now I find out that not only does this package not support 3.1.0 currently but I cannot seem to find any code generator that does :( big yikes! I have to figure out what my options are at this point as my spec currently makes heavy use of the `type` as an `array` not a `string` and I also use `oneOf` with `"null"` or a `$ref` throughout the whole thing which makes it a major problem to downgrade to `3.0.3` at this point.
> 
> I see that as of `27 days ago` at the time of me writing this, the 3.1 spec was added to [swagger-api/swagger-parser#1535](https://github.com/swagger-api/swagger-parser/issues/1535) - Does anyone have any idea when there will be a beta / snapshot release that will support 3.1.0 as if its going to be soon(tm), I might keep just waiting for that as an option before nuking my 3.1.0 setup....

Type null works with 3.0.3 specs and our tool
Type arrays can be converted to oneOf with each type in a different schema.

**Comment by philsturgeon** [comment]
Hello! As the maintainer of [openapi.tools](https://openapi.tools), and as somebody works with Linux Foundation helping out in OpenAPI-land, I'm reaching out to tooling vendors to track the progress towards supporting OpenAPI v3.1, to see what roadblocks there are beyond folks just generally being busy at this ridiculous time.

OpenAPI v3.1 has a bunch of great little changes, solving problems like the [JSON Schema <!=> OpenAPI Schema Object divergence](https://apisyouwonthate.com/blog/openapi-json-schema-divergence). It also fixes some other inconsistencies and duplicate ways of doing things. It's the best version and everyone should be using it, but we need tooling to catch up. Just in case folks didn't notice, or don't have resources to simplify the process, I'm here to give a friendly prod and send over some handy links.

Here are a few articles showing off the differences between OpenAPI v3.0 and v3.1.

- https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0
- https://www.apimatic.io/blog/2021/09/migrating-to-and-from-openapi-3-1/ 
- https://nordicapis.com/whats-new-in-openapi-3-1-0/

Here are some example files which can make for handy pass/fail test cases:

https://github.com/Mermade/openapi3-examples/tree/master/3.1 

If you're looking for the JSON Schema that defines a valid OpenAPI document, that'll be right over here: 

https://github.com/OAI/OpenAPI-Specification/tree/main/schemas/v3.1 


No rush, but when you're starting work on it, please update this issue so I can update [openapi.tools](https://openapi.tools) to reflect that, and folks will have a way to subscribe for updates.

LMK if you have any questions!

**Comment by marcelstoer** [comment]
Apart from the actual lack of 3.1 support the biggest issue I have is that this project does not seem to be open about it.

The [README](https://github.com/OpenAPITools/openapi-generator/blob/master/README.md) doesn't mention it. https://openapi-generator.tech/ prominently claims

> from OpenAPI 2.0/**3.x** documents

which led me to believe that 3.1 be fully supported.

**Comment by RickKukiela** [comment]
Probably should say 3.0.x...

**Comment by karenetheridge** [comment]
It would be helpful to see an itemized list of what is still needed to achieve v3.1 support, so that members of the community can help out. Has anyone identified those tasks?

**Comment by spacether** [comment]
> It would be helpful to see an itemized list of what is still needed to achieve v3.1 support, so that members of the community can help out. Has anyone identified those tasks?

I am putting together that list now and we will publish a project tracking it very soon. This is something that we all want to add to openapi-generator. It is ~30 feature additions to store the new openapi/json schema data in our java classes. Then that data needs to be used by the generators to generate code.

**Comment by spacether** [comment]
We have a project to track adding 3.1.0 support here:
https://github.com/orgs/OpenAPITools/projects/4/views/1 and welcome any PRs adding these features

**Comment by alexec** [comment]
Can we have a way to just treat v3.1 specs as v3.0? That way we could continue to use the generator with v3.1 APIs. Today we're blocked.

**Comment by spacether** [comment]
@alexec to do that one might need to preprocess the spec down to 3.0.3
If a 3.1.0 spec is parsed, it handles parsing differently and moves all type info into types.

**Comment by alexec** [comment]
Thank you for fast reply:

```yaml
components:
  schemas:
    Document:
      type: object
      properties:
        id:
          type: string
          example: abc123
        createdDate:
          type: string
          format: date-time
          example: 2023-01-01T00:00:00.000Z
        lastModifiedDate:
          type: string
          format: date-time
          example: 2023-01-01T00:00:00.000Z
        userName:
          type: string
          example: whoami
        content:
          type: string
          example: Hello APIs!
```

It creates this Java, but that's obviously wrong because the fields are `Object` when they should be `String`:

```java
public class Document {

  @JsonProperty("id")
  private Object id = null;

  @JsonProperty("createdDate")
  private Object createdDate = null;

  @JsonProperty("lastModifiedDate")
  private Object lastModifiedDate = null;

  @JsonProperty("userName")
  private Object userName = null;

  @JsonProperty("content")
  private Object content = null;
}
```

**Comment by spacether** [comment]
Right, like I said, the parser stores type info differently depending upon if the spec is 3.1.0 vs 3.0.3 or less

**Comment by pratik3558** [comment]
> Thank you for fast reply:
> 
> ```yaml
> components:
>   schemas:
>     Document:
>       type: object
>       properties:
>         id:
>           type: string
>           example: abc123
>         createdDate:
>           type: string
>           format: date-time
>           example: 2023-01-01T00:00:00.000Z
>         lastModifiedDate:
>           type: string
>           format: date-time
>           example: 2023-01-01T00:00:00.000Z
>         userName:
>           type: string
>           example: whoami
>         content:
>           type: string
>           example: Hello APIs!
> ```
> 
> It creates this Java, but that's obviously wrong because the fields are `Object` when they should be `String`:
> 
> ```java
> public class Document {
> 
>   @JsonProperty("id")
>   private Object id = null;
> 
>   @JsonProperty("createdDate")
>   private Object createdDate = null;
> 
>   @JsonProperty("lastModifiedDate")
>   private Object lastModifiedDate = null;
> 
>   @JsonProperty("userName")
>   private Object userName = null;
> 
>   @JsonProperty("content")
>   private Object content = null;
> }
> ```

Were you able to resolve the issue?
For me, its generating it as follows  when slotTypeId is a string:
`  private JsonNullable<Object> slotTypeId = JsonNullable.undefined();
`

**Comment by spacether** [comment]
For your info this type storage could all be way simpler long term. In my new version of this project types are defined [here](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/blob/2_0_0/modules/openapi-json-schema-generator/src/main/java/org/openapijsonschematools/codegen/model/CodegenSchema.java#L69)
I removed all the isString/Array etc properties and instead used:
```
    // 3.1.0
    public LinkedHashSet<String> types;
    public CodegenSchema contains;
    public LinkedHashMap<String, List<String>> dependentRequired;
    public boolean isBooleanSchemaTrue;  // supports boolean schemas
    public boolean isBooleanSchemaFalse;  // supports boolean schemas
```
which works for models and properties.
And to use it in templates, one can iterate over types like:
```
{{#each types}}
{{/each}}
```
Or one can check if types contains a value:
```
{{#contains types "string"}}
{{/contains}}
```

**Comment by GuillaumeSmaha** [comment]
> // 3.1.0
>     public LinkedHashSet<String> types;
>     public CodegenSchema contains;
>     public LinkedHashMap<String, List<String>> dependentRequired;
>     public boolean isBooleanSchemaTrue;  // supports boolean schemas
>     public boolean isBooleanSchemaFalse;  // supports boolean schemas

Oh, this is a better idea and it seems easier ! I will check your repo :)

**Comment by adampoplawski** [comment]
Hello
Is there a plan to add support for OpenAPI 3.1 in the 7.0.0 release?

**Comment by phillipuniverse** [comment]
Any sort of updated timeline anywhere? More and more upstream projects are supporting OpenAPI 3.1.

I see this project at https://github.com/orgs/OpenAPITools/projects/4, but seems like very little activity in the last year. And @spacether, what's your plan with your fork? I can't quite tell how complete the OpenAPI 3.1 support is - the README still seems to indicate OpenAPI 3.0.x only. And it seems like your fork is python generator-focused.

**Comment by phillipuniverse** [comment]
@spacether thanks for the extra information. Really unfortunate that some of the drama caused the fork! Your contributions speak for themselves! Seems like there is no other hope for OpenAPI 3.1 support landing in this repo 😢 

> in my v3 branch full statically check typing has been added to the python code which means that this generator could be ported into other static languages like Kotlin, Java, Go etc

Nice, maybe we could use the list at https://github.com/OpenAPITools/openapi-generator/issues/13073 for the next ones to prioritize. From this list, looks like it would be Java and typescript-axios.

**Comment by adampoplawski** [comment]
@spacether rhx for info. Are you planning to suport only Python in you fork in foreseeable future?

**Comment by spacether** [comment]
My focus is getting python working for 3.1.0 and as I said PRs and new generators in any language are welcome.

**Comment by abrahan-munro** [comment]
Note that if the spec document is mostly compatible with 3.0.x, just changing the version string from `3.1.0` to `3.0.3` can be used as a work around. When the version is 3.1, incorrect output is produced, but correct output may be generated if the version is 3.0. This is the case with the spec produced by the current version of FastAPI (0.100.0) for example.

**Comment by spacether** [comment]
Heads up, as of today just over 50% of the 3.1.0 new json schema keyword features have been implemented in my project: https://github.com/openapi-json-schema-tools/openapi-json-schema-generator
Including:
- const
- contains
- dependentRequired
- dependentSchemas
- maxContains
- minContains
- patternProperties
- prefixItems
- propertyNames
- unevaluatedItems
- unevaluatedProoerties

**Comment by Fares-Abubaker** [comment]
Hello all, 
when supporting for 3.1.0 will be released ?

**Comment by spacether** [comment]
> Hello! As the maintainer of [openapi.tools](https://openapi.tools), and as somebody works with Linux Foundation helping out in OpenAPI-land, I'm reaching out to tooling vendors to track the progress towards supporting OpenAPI v3.1, to see what roadblocks there are beyond folks just generally being busy at this ridiculous time.
> 
> ...
> LMK if you have any questions!

@philsturgeon one big roadblock is that there are a number of bugs in swagger-parser that prevent json schema tests cases  from running/passing. One can see 10 of those bugs here: https://github.com/swagger-api/swagger-parser/issues?q=is%3Aissue+is%3Aopen+3.1.0+author%3Aspacether
If people want these features to work, please consider helping fix these bugs in swagger-parser.

**Comment by philsturgeon** [comment]
@spacether I'k sorry to say I've got absolutely nothing to do with https://github.com/swagger-api/swagger-parser. You're mistaking it for https://github.com/APIDevTools/swagger-parser, which I do occasionally maintain, but it's is a whole other thing.

**Comment by spacether** [comment]
### OpenAPI v3.1.0 python code generation supported in [openapi-json-schema-generator](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator)
Heads up, I just released [v3.1.0 of openapi-json-schema-generator](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/releases/tag/3.1.0) which allows python code generation using a v3.1.0 spec.

Over 50% of the 3.1.0 new json schema keyword features have been implemented
Preliminary 3.1.0 spec support includes these json schema keywords:
1. type (array of types supported in addition to one non-array value)
2. const: only string values are working because of bugs in swagger parser
3. contains
4. dependentRequired
5. dependentSchemas
6. maxContains
7. minContains
8. patternProperties
9. prefixItems
10. propertyNames
11. unevaluatedItems
12. unevaluatedProperties

**Comment by spacether** [comment]
@adampoplawski @phillipuniverse I have been working on the Java generator and v3.0.0-v3.0.3 openapi schema validation has been implemented in java. This includes List/Map output classes, List/Map builder classes, and Enums which are accepted into Schema.validate and the List/Map builders. Read all about it [here](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator#reasons-to-use-the-java-generator). The Java schema validation will be available in my upcoming v4.0.0 release.

**Comment by sheldonreddy** [comment]
Hi, 

When can we expect the openapi specification version 3.1.0 to be supported by the OpenAPI-Generator tooling? Is there any roadmap we can reference to keep an eye on? 

Thanks

**Comment by wing328** [comment]
latest stable version v7.2.0 comes with beta support. you can give it a try to see if it works for your use cases.

**Comment by sheldonreddy** [comment]
Thanks @wing328 - will give it a shot.

**Comment by spacether** [comment]
@wing328 
- Where was the beta support added?
- What 3.1.0 features were added in what generators?
The only 3.1.0 PR that I see in that release is:
https://github.com/OpenAPITools/openapi-generator/pull/17174

**Comment by spacether** [comment]
## openapi v3.1.0 schema support added for java in openapi-json-schema-generator
Heads up, I just released [v4.1.0 of openapi-json-schema-generator](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/releases/tag/4.1.0) which allows java code generation using an openapi v3.1.0 document.

Over 66% of the 3.1.0 new json schema keyword features have been implemented
3.1.0 spec support includes these added/updated json schema keywords, which are tested with the json schema test suite in CI:
1. type (array of types supported in addition to one non-array value)
2. const: only string values are working because of bugs in swagger parser
3. contains
4. dependentRequired
5. dependentSchemas
6. maxContains
7. minContains
8. patternProperties
9. prefixItems
10. propertyNames
11. unevaluatedItems
12. unevaluatedProperties
13. if
14. then
15. else

**Comment by adampoplawski** [comment]
@wing328 Can you share the status of 3.1.0 support?

**Comment by wing328** [comment]
The best way to get updates regarding OpenAPI 3.1 support (beta) is via the Label in the pull request tab: https://github.com/OpenAPITools/openapi-generator/pulls?q=+is%3Apr+label%3A%22Feature%3A+OAS+3.1.0+spec+support%22+

**Comment by spacether** [comment]
# openapi 3.1.0 support added for the java generator in openapi-json-schema-generator
Heads up, [I just released v4.2.1 which includes the stable java sdk generator](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/releases/tag/4.2.1)
- It includes thorough endpoint code samples for schema validation an endpoint invocation [Money schema](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/blob/master/samples/client/petstore/java/docs/components/schemas/Money.md#money1) [Pet.addPet endpoint](https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/blob/master/samples/client/petstore/java/docs/apis/tags/Pet.md#addpet)
- Sealed classes are used for validated schema payloads/request bodies/responses/response bodies
- Code is re-used rather than generated multiple times: If an endpoint has multiple tags, the endpoint method is defined once as a default method in an interface, and that interface is implemented in each tag's api class

**Comment by jahlbornG** [comment]
@wing328 you mentioned that beta 3.1.0 support is available, but i can't see any information about enabling this support?  can you please provide more details?

**Comment by david0** [comment]
As complete OpenAPI 3.1 support doesn't seem to come soon, I would like to understand what the current state is. 

Wouldn't it make sense to document somewhere what experimental means and which features should working correctly?

**Comment by wing328** [comment]
@jahlbornG it's automatic. if your spec is documented as 3.1.0, it will enable the beta support. v7.7.0 was released last week. please give it a try as it includes more fixes for v3.1.0 spec

**Comment by david0** [comment]
The enabling is not the question, my question is what to expect? What does _beta_ mean? Should I expect all features to be there and just not as solid as 3.0?

**Comment by jahlbornG** [comment]
> @jahlbornG it's automatic. if your spec is documented as 3.1.0, it will enable the beta support. v7.7.0 was released last week. please give it a try as it includes more fixes for v3.1.0 spec

@wing328 i am using the generator maven plugin to convert a swagger document to an openapi spec document.  it seems to generate 3.0.1 by default, and i'd love to get it to generate 3.1.0.

**Comment by dankgen-tobias** [comment]
Is there a roadmap for the full support for 3.1? There is a project https://github.com/orgs/OpenAPITools/projects/4/views/1 that tackles this but there haven't been any changes since january

**Comment by trohovsky-mhp** [comment]
It would be great if openapi-generator tackled this issue https://github.com/OpenAPITools/openapi-generator/issues/9083

**Comment by Haarolean** [comment]
@wing328 what's the process of reporting issues encountered while building 3.1 specs? Should I raise a separate issue or sth else?

**Comment by wing328** [comment]
> Should I raise a separate issue

yes please

**Comment by bgalek** [comment]
Check out this poll: https://github.com/springdoc/springdoc-openapi/discussions/2790
I hope 3.1 will be default there, so this feature will be much appreciated ;)

**Comment by juliojgd** [comment]
Hi @bgalek we haven't found any relevant (**complete**) support nor (**REAL**) market adoption of OpenAPI 3.1 (I guess due is a wrong minor version, there are huge breaking changes and it should be a major version instead (among other potential reasons))

So what would be the point to move "official"  support to OpenAPI 3.1? Wouldn't it be better to wait for a proper and fixed major version (maybe OpenAPI 4.x)? And keep for the time being the "official" support to OpenAPI 3.0.x

**Comment by bgalek** [comment]
I've asked a few communities, and people do tend to miss some features in 3.0 (that were introduced to 3.1) like the major one I've described here: https://github.com/springdoc/springdoc-openapi/issues/2765#issuecomment-2476374042

After openapi@4 is published, tools will need to adjust and release new versions. That will take time, and it doesn't mean that 3.1 should not be supported, either. 

So, supporting 3.1 and then supporting upcoming 4 are separate tasks ;)

**Comment by juliojgd** [comment]
Supporting 3.1 I think will a be waste of resources that should be focused in any pending improvements in 3.0 support in all the related tools (OpenAPI generator, springdoc, etc.) and to speed up 4.0 support and adoption when it was released in the short future.

I think that concentrating the effort will be more positive than to disperse it in versions that were wrong from the inception  (3.1). To disperse the effort could lead to a less-than-optimal support in valuable versions that deserve the resource assignation.

**Comment by Stokestack** [comment]
Version 3.1 addresses several baffling defects in version 3.0 that cripple OAS for its intended purpose: documenting APIs.

One of the most significant fixes in 3.1 is the ability to provide a description alongside any use of a reference (model). This is critical in documenting an API, yet is inexplicably prohibited pre-3.1.

For example: I have a model that represents an integer-based setting on a piece of equipment. There will be dozens or more such settings in a complex API, all of them meaning something different. In every case, I need to describe what one particular integer-based setting is used for, as opposed to some other integer-based setting. Maybe I have one called "Volume" and another one called "Offset" and another called "Tap count." Every one of those is an _Integer_setting_. How can you document an API without describing each use of its structures?

One can think of endless examples. What about an API that had something to do with surveying or farming, where you needed to work with rectangular land parcels. You might have a model called Rect, referred-to all over the place. What does this Rect mean as opposed to that one? You need to be able to attach a description to each appearance of Rect in the API.

Versions before 3.1 are crippled by this and other design problems. To continue to let it languish, unusable, for many purposes until a nebulous version 4 appears... and for several more years after that until tools actually support it... seems absurd. You might as well just call OpenAPI a wrap and move on to something else.

**Comment by darrelmiller** [comment]
There is definitely support for OpenAPI 3.1 in tools, and we are starting to see more 3.1 descriptions in the wild.  You can see a list here https://openapi.tools/ of tools that declare they support 3.1.  It does take a long time for new versions of specifications to take hold.  We saw this in the 2.0 -> 3.0 transition.

Adding 3.1 support for code generators is harder than it is for most tooling because 3.1 opened the doors to a range of different capabilities in JSON Schema 2020-12.

@juliojgd  It is certainly debatable as to whether 3.1 should have been 4.0.  And certainly to some implementors it feels that way.  However, to authors of API Descriptions the changes are 98% additive and not breaking.  The decision was hotly debated and it came down to a vote.  So, be assured that if you think it should have been 4.0, there are some members of the OpenAPI technical steering committee that agree with you. 

But here we are with a 3.1 that is easy for some to support and hard for others.  But it is an important step to make, and one that needed to be made.  For those saying you should wait for 4.0.  I wouldn't recommend that.  4.0 will be a major update and as we have explored the features we want to bring, we have identified that many of the improvements would not be breaking.  So, rather than make people wait, for those enhancements, we will be continuing the 3.x line while working on the 4.0 release.  There will be a 3.2 release "fairly soon" that will add some simple to add features that should be easy to support.  However, we currently are planning a 3.3. release to follow to backport some of the ideas we came up with for moonwalk (4.0).

While this may seem like a lot of work to do for a group that hasn't released a new version in 3 years, we have recently grown the technical steering committee, and there is lots of new life in the group.  

@Stokestack I'm glad you like the ability to add description and summary alongside $refs.  This previous constraint wasn't a decision made the OpenAPI group. It was a constraint of earlier versions of JSON Schemas because the semantics of merging keywords was not clearly defined.  By adopting JSON Schema 2020-12, were able to leverage this work and start bringing peer keywords.  We only started with "description" and "summary" because we knew they were safe.  If there are other keywords that you believe would be valuable to sit alongside a $ref, let us know by creating an issue in the OpenAPI specification repo.

**Comment by Stokestack** [comment]
@darrelmiller Thanks for that response. I don't understand how JSON Schema precluded anything, but I haven't studied it. But I didn't think OAS 3 was JSON-Schema-compliant anyway, so I'm perplexed as to why it would be constrained by failings of that standard.

I will keep your suggestion in mind about suggesting other keywords.

**Comment by wing328** [comment]
@spacether work can be found in https://github.com/openapitools/openapi-json-schema-generator and he decided to host it elsewhere instead. His approach unfortunately didn't gain attraction and many Python users we contacted switched to something else instead, and therefore we refactored the Python client generator (using pydantic) and have [gained many positive feedback](https://github.com/OpenAPITools/openapi-generator/pull/14157) and [enhancements](https://github.com/OpenAPITools/openapi-generator/pulls?page=6&q=is%3Apr+is%3Amerged+label%3A%22Client%3A+Python%22) since the beta release.

As said [before](https://github.com/OpenAPITools/openapi-generator/issues/9083#issuecomment-2009077256), the best way to get updates regarding OpenAPI 3.1 support (beta) is via the Label in the pull request tab: https://github.com/OpenAPITools/openapi-generator/pulls?q=+is%3Apr+label%3A%22Feature%3A+OAS+3.1.0+spec+support%22+, and [raise a new issue if needed](https://github.com/OpenAPITools/openapi-generator/issues/9083#issuecomment-2301420928)

For general feedback on OpenAPI specification, please open an issue in the [official repo](https://github.com/OAI/OpenAPI-Specification) as this project is not affiliated with OpenAPI Initiative (OAI).
